pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,265 @@
+// Shards and APIExports are read from the cache server
+func NewController(
+	apiExportEndpointSliceClusterInformer apisinformers.APIExportEndpointSliceClusterInformer,
+	localShardClusterInformer corev1alpha1informers.ShardClusterInformer,
+	globalShardClusterInformer corev1alpha1informers.ShardClusterInformer,
+	localAPIExportClusterInformer apisinformers.APIExportClusterInformer,
+	globalAPIExportClusterInformer apisinformers.APIExportClusterInformer,
+	kcpClusterClient kcpclientset.ClusterInterface,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+		listAPIExportEndpointSlices: func() ([]*apisv1alpha1.APIExportEndpointSlice, error) {
+			return apiExportEndpointSliceClusterInformer.Lister().List(labels.Everything())
+		},
+		listShards: func() ([]*corev1alpha1.Shard, error) {
+			shards, err := globalShardClusterInformer.Lister().List(labels.Everything())
+			if err != nil {
+				var localErr error
+				shards, localErr = localShardClusterInformer.Lister().List(labels.Everything())
+				if localErr != nil {
+					err = localErr
+				}
+			}
+			return shards, err
+
+		},
+		getAPIExportEndpointSlice: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExportEndpointSlice, error) {
+			return apiExportEndpointSliceClusterInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		getAPIExport: func(clusterName logicalcluster.Name, path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error) {
+			apiExport, err := localAPIExportClusterInformer.Lister().Cluster(clusterName).Get(name)
+			if errors.IsNotFound(err) {
+				return indexers.ByPathAndName[*apisv1alpha1.APIExport](apisv1alpha1.Resource("apiexports"), globalAPIExportClusterInformer.Informer().GetIndexer(), path, name)
+			}
+			return apiExport, err
+		},
+		commit: committer.NewCommitter[*APIExportEndpointSlice, Patcher, *APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus](kcpClusterClient.ApisV1alpha1().APIExportEndpointSlices()),
+	}
+
+	indexers.AddIfNotPresentOrDie(globalAPIExportClusterInformer.Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPathAndName: indexers.IndexByLogicalClusterPathAndName,
+	})
+
+	apiExportEndpointSliceClusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExportEndpointSlice(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+	})
+
+	globalShardClusterInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(newObj)
```

Comment:
do you know how frequent a `shard` object is updated ?

Reply:
since we only care about the URL address we could simply compare the old and the new objs and see if it makes sense to trigger the reconciliation

Reply:
In the next PR we care about the labels as well

Reply:
it seems that this change doesn't require a lot. I think it it a 3 liner, no?

Reply:
would rather like to see a filtered handler that only handles specific updates.